File: library/src/main/java/com/cyb3rko/abouticons/AboutIcons.java
package com.cyb3rko.abouticons;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import es.dmoral.toasty.Toasty;

public class AboutIcons {
    private ArrayList<IconModel> modelList = new ArrayList<>();
    private Class drawableClass;
    private Context appContext;
    private RecyclerView recyclerView;
    private RecyclerViewAdapter mAdapter;
    private View view;

    private boolean allowModificationAnnotation = true;

    @SuppressLint("InflateParams")
    public AboutIcons(Context appContext, Class drawableClass) {
        this.appContext = appContext;
        this.drawableClass = drawableClass;

        Toasty.Config.getInstance().allowQueue(false).apply();
        view = LayoutInflater.from(appContext).inflate(R.layout.activity_icon_view, null);
        recyclerView = view.findViewById(R.id.recycler_view);
    }

    private void setAdapter() {
        mAdapter = new RecyclerViewAdapter(appContext, modelList, drawableClass, allowModificationAnnotation);

        for (int i = 0; i < mAdapter.getIconListSize(); i++) {
            addAttributes(i);
        }

        final GridLayoutManager layoutManager = new GridLayoutManager(appContext, 2);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(mAdapter);

        onItemClick();
    }

    private void addAttributes(int index) {
        boolean modified = false;
        String iconAuthor = "[Missing]";
        String iconWebsite = "[Missing]";
        String iconLink = "";

        try {
            int arrayId = appContext.getResources().getIdentifier(mAdapter.getDrawableName(index), "array", appContext.getPackageName());
            String[] iconInformation = appContext.getResources().getStringArray(arrayId);
            modified = Boolean.parseBoolean(iconInformation[3]);
            iconAuthor = "by " + iconInformation[0];
            iconWebsite = iconInformation[1];
            iconLink = iconInformation[2];
        } catch (Exception ignored) {
        }

        modelList.add(new IconModel(modified, iconAuthor, iconWebsite, iconLink));
    }

    private void onItemClick() {
        mAdapter.setOnItemClickListener(new RecyclerViewAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position, IconModel model) {
                if (!model.getIconLink().equals("")) {
                    appContext.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(model.getIconLink())));
                } else {
                    Toasty.warning(appContext, "Link not set!", Toasty.LENGTH_SHORT).show();
                }
            }
        });
    }

    public AboutIcons hideTitle() {
        TextView titleView = view.findViewById(R.id.title_view);
        titleView.setVisibility(View.GONE);
        return this;
    }

    public AboutIcons setTitle(String customTitle) {
        TextView titleView = view.findViewById(R.id.title_view);
        titleView.setText(customTitle);
        return this;
    }

    public AboutIcons hideModificationAnnotation() {
        view.findViewById(R.id.modified_info).setVisibility(View.GONE);
        allowModificationAnnotation = false;
        return this;
    }

    public View get() {
        setAdapter();
        return view;
    }
}

File: library/src/main/java/com/cyb3rko/abouticons/AboutIcons.kt
package com.cyb3rko.abouticons

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.*
import java.util.*

class AboutIcons(
    private val appContext: Context,
    private val drawableClass: Class<*>,
    private val fragmentManager: FragmentManager
) {

    private var allowModificationAnnotation: Boolean = true
    private lateinit var mAdapter: RecyclerViewAdapter
    private val modelList = ArrayList<IconModel>()
    private val modifiedContainer: LinearLayout
    private var progressBar: ProgressBar
    private var recyclerView: RecyclerView
    private var titleView: TextView
    private var view: View

    init {
        @SuppressLint("InflateParams")
        view = LayoutInflater.from(appContext).inflate(R.layout.activity_icon_view, null)
        recyclerView = view.findViewById(R.id.recycler_view)
        modifiedContainer = view.findViewById(R.id.modified_container)
        progressBar = view.findViewById(R.id.progress_bar)
        titleView = view.findViewById(R.id.title_view)
    }

    private fun setAdapter() {
        val layoutManager = GridLayoutManager(appContext, 2)
        recyclerView.layoutManager = layoutManager

        CoroutineScope(Dispatchers.Main).launch {
            mAdapter = RecyclerViewAdapter(appContext, modelList, drawableClass, allowModificationAnnotation)
            for (i in 0 until mAdapter.getIconListSize()) {
                val task = async { fetchAttributes(i) }
                modelList.add(task.await())
            }

            recyclerView.adapter = mAdapter
            onItemClick()
            progressBar.visibility = View.GONE
        }
    }

    private fun fetchAttributes(index: Int): IconModel {
        val missingHint = "[${appContext.getString(R.string.missing)}]"
        var iconAuthor = missingHint
        var iconWebsite = missingHint
        var iconLink = ""
        var modified = false
        var iconLicense = ""

        try {
            val arrayId = appContext.resources.getIdentifier(mAdapter.getDrawableName(index), "array", appContext.packageName)
            val iconInformation = appContext.resources.getStringArray(arrayId)
            iconAuthor = iconInformation[0]
            iconWebsite = iconInformation[1]
            iconLink = iconInformation[2]
            modified = iconInformation[3]!!.toBoolean()
            iconLicense = iconInformation[4]
        } catch (ignored: Exception) {
        }

        return IconModel(iconAuthor, iconWebsite, iconLink, modified, iconLicense)
    }

    private fun onItemClick() {
        mAdapter.setOnItemClickListener(object: RecyclerViewAdapter.OnItemClickListener {
            override fun onItemClick(view: View?, position: Int, model: IconModel?) {
                if (model != null) {
                    IconInfoBuilder(appContext, fragmentManager)
                        .setDrawable(mAdapter.getDrawableInt(model))
                        .setLink(model.iconLink)
                        .setAuthor(model.author)
                        .setWebsite(model.website)
                        .setModifiedInfo(model.modified)
                        .setLicense(model.iconLicense)
                        .start()
                }
            }
        })
    }

    fun setTitle(customTitle: String): AboutIcons {
        if (customTitle.isNotBlank()) {
            titleView.text = customTitle
        } else {
            titleView.visibility = View.GONE
        }
        return this
    }

    fun setTitleSize(customSize: Float): AboutIcons {
        titleView.textSize = customSize
        return this
    }

    fun setModificationTitle(customTitle: String): AboutIcons {
        view.findViewById<TextView>(R.id.modified_text).text = customTitle
        return this
    }

    fun setModificationTitleSize(customSize: Float): AboutIcons {
        view.findViewById<TextView>(R.id.modified_text).textSize = customSize
        return this
    }

    fun hideModificationAnnotation(): AboutIcons {
        modifiedContainer.visibility = View.GONE
        allowModificationAnnotation = false
        return this
    }

    fun get(): View {
        setAdapter()
        if (titleView.visibility == View.GONE && modifiedContainer.visibility == View.GONE) {
            view.findViewById<CardView>(R.id.header).visibility = View.GONE
        } else {
            view.findViewById<ImageView>(R.id.color_sample).colorFilter = PorterDuffColorFilter(ContextCompat.getColor(appContext,
                R.color.colorModified), PorterDuff.Mode.SRC_ATOP)
        }
        return view
    }

    @Deprecated(message = "Instead use setTitle(\"\")", replaceWith = ReplaceWith("this.setTitle(\"\")"))
    fun hideTitle(): AboutIcons {
        val titleView: TextView = view.findViewById(R.id.title_view)
        titleView.visibility = View.GONE
        return this
    }
}
